finite_approx/data_set_lib.py

```diff
@@ -114,3 +114,60 @@ def run_newton_tr(self, params_init, maxiter=200, gtol=1e-6):
 
         print('Done with Newton trust region.')
         return tr_opt
+
+    def get_log_q_pi(self, params, pi):
+        tau, phi_mu, phi_var, nu = self.unpack_params(params)
+        return log_q_pi(pi, tau)
+
+    def get_variational_log_lh(self, params, a, z, pi):
+        tau, phi_mu, phi_var, nu = self.unpack_params(params)
+        return log_q_a(a, phi_mu.T, phi_var) + log_q_z(z, nu) + log_q_pi(pi, tau)
+
+
+###################
+# Variational likelihoods
+
+# normal means
+def log_q_a(a, phi_mu, phi_var):
+    # note the shapes below: phi_mu should be the "correct",
+    # not the transposed shape
+    k_approx = np.shape(phi_mu)[0]
+    x_d = np.shape(phi_mu)[1]
+    assert np.shape(phi_mu)[0] == len(phi_var), 'shape of phi_var and phi_mu do not match'
+    assert np.shape(a) == np.shape(phi_mu), 'shape of A and phi_mu do not match'
+
+    log_denom = (x_d/2) * np.sum(np.log(2 * np.pi * phi_var))
+    return -0.5 * np.sum((a.T - phi_mu.T) ** 2 / phi_var) - log_denom
+
+# bernoulli responsibilities
+def log_q_z(z, nu):
+    return np.sum(np.log(nu**z) + np.log((1 - nu)**(1-z)))
+
+# pi stick lengths
+def log_q_pi(pi, tau):
+    log_beta = sp.special.gammaln(tau[:,0]) + sp.special.gammaln(tau[:,1]) \
+                    - sp.special.gammaln(tau[:,0] + tau[:,1])
+
+    return np.sum(-log_beta + (tau[:,0] - 1) * np.log(pi) \
+                    + (tau[:,1] - 1) * np.log(1 - pi))
+
+####################
+# Prior likelihoods
+def log_p0_a(a, sigma_a):
+    prior_mean = np.zeros(np.shape(a))
+    prior_var = np.ones(np.shape(a)[0]) * sigma_a
+    return log_q_a(a, prior_mean, prior_var)
+
+# bernoulli responsibilities
+def log_p0_z(z, pi):
```

Comment:
I'm not 100% sure what you have in mind for this function.  I think of P(Z | pi) as part of the likelihood, not a prior, but semantics aside, if you want to calculate sensitivity WRT the functional specification of P(Z | pi), you'll need E_q [ log P(Z | pi) ] I think.

Maybe I'm forgetting something.  Unfortunately in github it's not easy to see notebooks -- having left this comment, I'll check this out and try to see for myself how you're using it.